tests/cpp/test_multidevice_sharding.cpp

```diff
@@ -807,4 +809,246 @@ TEST_F(MultiDeviceTest, ShardedSplitReshapeIds) {
+    std::vector<TensorView*> output_tvs;
+    for (auto output : expr->outputs()) {
+      output_tvs.push_back(output->as<TensorView>());
+    }
+
+    TransformPropagator propagator(input_tv);
+
+    // Note: We will finally propagate from each input iteratively.
+    SetSelector selector(
+        std::unordered_set<TensorView*>(output_tvs.begin(), output_tvs.end()));
+    MaxLogicalDomainInfoSpanningTree(input_tv, &selector).traverse(&propagator);
+    scheduler_utils::parallelizeAllLike(
+        input_tv,
+        /*pos=*/-1,
+        /*selected_tv=*/output_tvs);
+  }
+}
+
+} // namespace
+
+TEST_F(MultiDeviceTest, TransformerFwd) {
+  auto fusion = std::make_unique<Fusion>();
+  FusionGuard fg(fusion.get());
+
+  const int d = communicator_->size();
+  const int64_t b = 2, s = 3, h = 8, e = 16;
+  auto mesh = DeviceMesh::createForNumDevices(d);
+
+  std::vector<int64_t> in_shape = {b, s, d * h * e};
+  std::vector<int64_t> out_shape = {b, s, d * h, e};
+
+  // The transformer block produces hq/hk/hv after slicing the MHA linear
+  // output.
+  TensorView* hq = makeConcreteTensor(in_shape, DataType::Half);
+  TensorView* hk = makeConcreteTensor(in_shape, DataType::Half);
+  TensorView* hv = makeConcreteTensor(in_shape, DataType::Half);
+
+  TensorView* q = reshape(hq, in_shape, out_shape);
+  TensorView* q_permuted = permute(q, {0, 2, 1, 3});
+  TensorView* k = reshape(hk, in_shape, out_shape);
+  TensorView* k_permuted = permute(k, {0, 2, 1, 3});
+  TensorView* v = reshape(hv, in_shape, out_shape);
+  TensorView* v_permuted = permute(v, {0, 2, 1, 3});
+
+  SdpfaFwdResult sdpa_out = sdpfa_fwd(
+      q_permuted,
+      k_permuted,
+      v_permuted,
+      /*dropout_p=*/IrBuilder::create<Val>(0.0),
+      /*is_causal=*/IrBuilder::create<Val>(false),
+      /*scale=*/nullptr);
+
+  TensorView* attn = sdpa_out.output;
+  TensorView* attn_permute = permute(attn, {0, 2, 1, 3});
+  TensorView* out = reshape(attn_permute, out_shape, in_shape);
+
+  fusion->addInput(hq);
+  fusion->addInput(hk);
+  fusion->addInput(hv);
+  fusion->addOutput(out);
+
+  // Shard input tensors
+  for (auto* tv : {hq, hk, hv}) {
+    tv->setDeviceMesh(mesh);
+    tv->split(-1, d, /*inner_split=*/false);
+    tv->axis(-2)->parallelize(ParallelType::DIDx);
+    reorderDIDToFront(tv);
+  }
+  propagateShardings(fusion.get(), d);
+
+  for (auto tv : fusion->allTvs()) {
+    tv->setAllocationDomain(tv->getLoopDomain(), true);
+  }
+
+  FusionExecutorCache executor_cache(std::move(fusion));
+  at::Tensor hq_tensor = at::randn({in_shape}, tensor_options.dtype(at::kHalf));
+  at::Tensor hk_tensor = at::randn({in_shape}, tensor_options.dtype(at::kHalf));
+  at::Tensor hv_tensor = at::randn({in_shape}, tensor_options.dtype(at::kHalf));
+
+  at::Tensor sharded_hq = shardTensor(hq_tensor, -1, mesh);
+  at::Tensor sharded_hk = shardTensor(hk_tensor, -1, mesh);
+  at::Tensor sharded_hv = shardTensor(hv_tensor, -1, mesh);
+
+  at::Tensor nvf_out =
```

Comment:
```suggestion
  auto nvf_out =
```
the `.as` is clear about the type